### ISSUE TITLE
Gitea as an OpenID-2.0 provider

### ISSUE THREAD

**Issue by strk** [go-gitea/gitea#1310]
Now that OpenID instances can be configured to accept an OpenID login, it could be nice if they ccould also serve as OpenID providers so to build what would become the authentication layer of a federation of code repositories. See also #184

**Comment by cweiske** [comment]
Why is every software trying to become a OpenID provider?

I do not think that it's necessary for federated pull requests to be an OpenID provider.

**Comment by jhasse** [comment]
It isn't necessary, but it makes it easier as you don't have to create a traditional account on every Gitea instance where you want to open a pull request (and you don't have or don't want to use a different OpenID provider).

**Comment by bkcsoft** [comment]
If you can setup your own gitea-instance you can setup your own OpenID provider. I _really_ don't get why Gitea has to be a provider for everything 😒

**Comment by MaxG87** [comment]
Albeit this thread is quiet for some time I want to add some points for the sake of the argument.

> If you can setup your own gitea-instance you can setup your own OpenID provider.

And instantly you have to maintain two systems instead of one. Because a project decided not to contribute to infrastructure it partly relies on. And maintaining more systems in a self-hosting setting means much more opportunity to miss important updates.

Furthermore,  at least for me, the condition of free, decentralised OpenID infrastructure seems to be quite bad. You can reuse the logins of your accounts at the usual tech giants, but only because someone hardcoded the support for them. I hardly know a service where you can use OpenID tokens from the provider of your choice. I also don't know providers of such tokens that could be used freely. All I know off is the possibility to integrate third-party-logins in your website for preselected third parties.

Long story short: OpenID needs more providers, otherwise Giteas OpenID feature cannot be used in a decentralised way. Gitea should be such a provider to reduce required mainenance efforts in self-hosting settings.

> I _really_ don't get why Gitea has to be a provider for everything

It is only about OpenID, isn't it? For me, this seems to be much less than everything.

**Comment by lunny** [comment]
Since Gitea now could be as OAuth2 provider, that's not too difficult.

**Comment by strk** [comment]
I still think this is an important feature. Since Gitea does support `OpenID-2.0`, by acting as a provider would allow accessing all enabled Gitea instances via identity provided by your own. A big step toward federation.

**Comment by zeripath** [comment]
If you put links to the specs in this issue anyone that's interested in building this functionality will have an easier time doing it.

**Comment by 6543** [comment]
gitea now can act as openID provider - so close this issue?

**Comment by lunny** [comment]
Which PR resolved this?

**Comment by 6543** [comment]
https://docs.gitea.io/en-us/oauth2-provider/?

**Comment by jolheiser** [comment]
> https://docs.gitea.io/en-us/oauth2-provider/?

OpenID is not OAuth2  

For example, OAuth2 allows another service to access the Gitea API on your behalf, whereas OpenID would allow another service to treat your Gitea account as a valid account on their platform.

**Comment by philpraxis** [comment]
That would be great if Gitea can become OpenID Connect provider !
I don't know if I can ask, but let's go: Can someone familiar with the code explain the level of complexity (or days / weeks / months dev time) that such feature would need?

**Comment by strk** [comment]
I think we want discovery for this ticket to be closed. Without discovery a Gitea instance could not be used as an identity provider for other Gitea instances. This ticket was about `OpenID-2.0` for this reason (decentralization of identities)

**Comment by strk** [comment]
Why was this closed ? There's no discovery implemented.

**Comment by 6543** [comment]
@strk I think so: https://try.gitea.io/.well-known/openid-configuration

**Comment by strk** [comment]
But that's for oath, not OpenID-2.0, right ?

**Comment by 6543** [comment]
hmm I thought it is for OpenID, to auto-discover things ...

**Comment by cweiske** [comment]
That is OpenID connect. OpenID 2 is something completely different :(

**Comment by jtackaberry** [comment]
This feature becomes even more important with Gitea Actions.  For example, by adding one's Gitea instance as an IdP in AWS IAM, it would allow jobs to assume IAM roles without the need to store AWS credentials as secrets (and the associated burden of regular rotations).

GitHub can do this: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers
GitLab too, FWIW: https://docs.gitlab.com/ee/ci/cloud_services/aws/

GitHub requires jobs to hit an endpoint defined in `$ACTIONS_ID_TOKEN_REQUEST_URL` to retrieve a token (typically via an action -- [for example](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) -- though it can be done with curl as shown in the above URL), while GitLab automatically generates one and injects it into the job's environment.  I find GitLab's approach much more convenient, but I suppose Gitea would follow GitHub's approach given it's copying Actions.
